docs/source/configuration/entity-caching.mdx

```diff
@@ -129,9 +128,162 @@ This entry contains an object with the `all` field to affect all subgraph reques
+- All cached entries for a specific entity in a specific subgraph
+
+To process an invalidation request, the router first sends a `SCAN` command to Redis to find all the keys that match the invalidation request. After iterating over the scan cursor, the router sends a `DEL` command to Redis to remove the matching keys.
+
+#### Configuration
+
+You can configure entity cache invalidation both globally (`preview_entity_cache.subgraph.all.invalidation`) and per subgraph (`preview_entity_cache.subgraph.subgraphs.invalidation`). 
+
+```yaml title="router.yaml"
+preview_entity_cache:
+  enabled: true
+  redis:
+    urls: ["redis://..."]
+
+  subgraph:
+    all:
+      enabled: true
+      # global invalidation configuration
+      invalidation:
+        # address of the invalidation endpoint
+        # this should only be exposed to internal networks
+        listen: "127.0.0.1:3000"
+        path: "/invalidation"
+        # base64 string that will be provided in the `Authorization: Basic` header value
+        shared_key: "agm3ipv7egb78dmxzv0gr5q0t5l6qs37"
+    subgraphs:
+      products:
+        # per subgraph invalidation configuration
+        invalidation:
+          # whether invalidation is enabled for this subgraph
+          enabled: true
+          # override the shared key for this particular subgraph. If another key is provided, the invalidation requests for this subgraph's entities will not be executed
+          shared_key: "czn5qvjylm231m90hu00hgsuayhyhgjv"
+```
+
+#### Invalidation request format
+
+Invalidation requests are sent as JSON objects with the following format:
+
+- Subgraph invalidation request:
+
+```json
+{
+  "kind": "subgraph",
+  "subgraph": "accounts"
+}
+```
+
+- Subgraph type invalidation request:
+
+```json
+{
+  "kind": "subgraph",
+  "subgraph": "accounts",
+  "type": "User"
+}
+```
+
+- Subgraph entity invalidation request:
+
+```json
+{
+  "kind": "subgraph",
+  "subgraph": "accounts",
+  "type": "User",
+  "key": {
+    "id": "1"
+  }
+}
+```
+
+<Note>
+
+This is the same key format used by the router in the `representations` field for `_entities` queries.
+
+</Note>
+
+
+#### Invalidation HTTP endpoint
+
+The invalidation endpoint exposed by the router can receive multiple invalidation requests at once and will process them in sequence. For authorization, you can provide a shared key in the request. For example, with the previous configuration you can send the following request:
+
+```
+POST http://127.0.0.1:3000/invalidation
+Authorization: agm3ipv7egb78dmxzv0gr5q0t5l6qs37
+Content-Length:96
+Content-Type:application/json
+Accept: application/json
+
+[{
+    "kind": "type",
+    "subgraph": "invalidation-subgraph-type-accounts",
+    "type": "Query"
+}]
+```
+
+We would expect the following response:
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "count": 300
+}
+```
+
+The `count` field indicates the number of keys that were removed from Redis.
```

Comment:
It's possibly a technicality, but we haven't implemented this. The invalidation `delete` function actually uses the number of matching keys from the scan. The underlying `RedisCacheStorage` `delete` function returns an `Option` which does contain the number of deleted keys, so we could fix that as something like:
```
-                        count += keys.len() as u64;
-                        storage.delete(keys).await;
+                        count += storage.delete(keys).await.unwrap_or(0) as u64;
```

Reply:
that change has been merged now